src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKeyManager.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal class AccessKeyManager : IAccessKeyManager, IDisposable
+    {
+        private readonly TimerAwaitable _timer = new TimerAwaitable(TimeSpan.Zero, TimeSpan.FromMinutes(5));
+
+        private readonly ConcurrentDictionary<ServiceEndpoint, object> _endpoints = new ConcurrentDictionary<ServiceEndpoint, object>();
+
+        public AccessKeyManager(
+            IServerNameProvider nameProvider,
+            ILoggerFactory loggerFactory
+            )
+        {
+            _ = UpdateAsync(nameProvider, loggerFactory);
+        }
+
+        private async Task UpdateAsync(IServerNameProvider provider, ILoggerFactory factory)
+        {
+            using (_timer)
+            {
+                _timer.Start();
+
+                while (await _timer)
+                {
+                    var keys = new HashSet<AadAccessKey>();
```

Comment:
Should `AadAccessKey` for each `ServiceEndpoint` be different from each other?  If so, `HashSet` is not needed. Or you can just store the `AadAccessKey` in the `ConcurrentDictionary` with a proper `AadAccessKey` equality comparer.